File: mem0/vector_stores/pgvector.py
import json
import logging
from typing import List, Optional

from pydantic import BaseModel

# Try to import psycopg (psycopg3) first, then fall back to psycopg2
try:
    import psycopg
    from psycopg import execute_values
    from psycopg.types.json import Json
    PSYCOPG_VERSION = 3
    logger = logging.getLogger(__name__)
    logger.info("Using psycopg (psycopg3) for PostgreSQL connections")
except ImportError:
    try:
        import psycopg2
        from psycopg2.extras import execute_values, Json
        PSYCOPG_VERSION = 2
        logger = logging.getLogger(__name__)
        logger.info("Using psycopg2 for PostgreSQL connections")
    except ImportError:
        raise ImportError(
            "Neither 'psycopg' nor 'psycopg2' library is available. "
            "Please install one of them using 'pip install psycopg' or 'pip install psycopg2'."
        )

from mem0.vector_stores.base import VectorStoreBase

logger = logging.getLogger(__name__)


class OutputData(BaseModel):
    id: Optional[str]
    score: Optional[float]
    payload: Optional[dict]


class PGVector(VectorStoreBase):
    def __init__(
        self,
        dbname,
        collection_name,
        embedding_model_dims,
        user,
        password,
        host,
        port,
        diskann,
        hnsw,
        sslmode=None,
        connection_string=None,
        connection_pool=None,
    ):
        """
        Initialize the PGVector database.

        Args:
            dbname (str): Database name
            collection_name (str): Collection name
            embedding_model_dims (int): Dimension of the embedding vector
            user (str): Database user
            password (str): Database password
            host (str, optional): Database host
            port (int, optional): Database port
            diskann (bool, optional): Use DiskANN for faster search
            hnsw (bool, optional): Use HNSW for faster search
            sslmode (str, optional): SSL mode for PostgreSQL connection (e.g., 'require', 'prefer', 'disable')
            connection_string (str, optional): PostgreSQL connection string (overrides individual connection parameters)
            connection_pool (Any, optional): psycopg2 connection pool object (overrides connection string and individual parameters)
        """
        self.collection_name = collection_name
        self.use_diskann = diskann
        self.use_hnsw = hnsw
        self.embedding_model_dims = embedding_model_dims

        # Connection setup with priority: connection_pool > connection_string > individual parameters
        if connection_pool is not None:
            # Use provided connection pool
            self.conn = connection_pool.getconn()
            self.connection_pool = connection_pool
        elif connection_string is not None:
            # Use connection string
            if sslmode:
                # Append sslmode to connection string if provided
                if 'sslmode=' in connection_string:
                    # Replace existing sslmode
                    import re
                    connection_string = re.sub(r'sslmode=[^ ]*', f'sslmode={sslmode}', connection_string)
                else:
                    # Add sslmode to connection string
                    connection_string = f"{connection_string} sslmode={sslmode}"
            
            if PSYCOPG_VERSION == 3:
                self.conn = psycopg.connect(connection_string)
            else:
                self.conn = psycopg2.connect(connection_string)
            self.connection_pool = None
        else:
            # Use individual connection parameters
            conn_params = {
                'dbname': dbname,
                'user': user,
                'password': password,
                'host': host,
                'port': port
            }
            if sslmode:
                conn_params['sslmode'] = sslmode
            
            if PSYCOPG_VERSION == 3:
                self.conn = psycopg.connect(**conn_params)
            else:
                self.conn = psycopg2.connect(**conn_params)
            self.connection_pool = None
        
        self.cur = self.conn.cursor()

        collections = self.list_cols()
        if collection_name not in collections:
            self.create_col(embedding_model_dims)

    def create_col(self, embedding_model_dims):
        """
        Create a new collection (table in PostgreSQL).
        Will also initialize vector search index if specified.

        Args:
            embedding_model_dims (int): Dimension of the embedding vector.
        """
        self.cur.execute("CREATE EXTENSION IF NOT EXISTS vector")
        self.cur.execute(
            f"""
            CREATE TABLE IF NOT EXISTS {self.collection_name} (
                id UUID PRIMARY KEY,
                vector vector({embedding_model_dims}),
                payload JSONB
            );
        """
        )

        if self.use_diskann and embedding_model_dims < 2000:
            # Check if vectorscale extension is installed
            self.cur.execute("SELECT * FROM pg_extension WHERE extname = 'vectorscale'")
            if self.cur.fetchone():
                # Create DiskANN index if extension is installed for faster search
                self.cur.execute(
                    f"""
                    CREATE INDEX IF NOT EXISTS {self.collection_name}_diskann_idx
                    ON {self.collection_name}
                    USING diskann (vector);
                """
                )
        elif self.use_hnsw:
            self.cur.execute(
                f"""
                CREATE INDEX IF NOT EXISTS {self.collection_name}_hnsw_idx
                ON {self.collection_name}
                USING hnsw (vector vector_cosine_ops)
            """
            )

        self.conn.commit()

    def insert(self, vectors, payloads=None, ids=None):
        """
        Insert vectors into a collection.

        Args:
            vectors (List[List[float]]): List of vectors to insert.
            payloads (List[Dict], optional): List of payloads corresponding to vectors.
            ids (List[str], optional): List of IDs corresponding to vectors.
        """
        logger.info(f"Inserting {len(vectors)} vectors into collection {self.collection_name}")
        json_payloads = [json.dumps(payload) for payload in payloads]

        data = [(id, vector, payload) for id, vector, payload in zip(ids, vectors, json_payloads)]
        execute_values(
            self.cur,
            f"INSERT INTO {self.collection_name} (id, vector, payload) VALUES %s",
            data,
        )
        self.conn.commit()

    def search(self, query, vectors, limit=5, filters=None):
        """
        Search for similar vectors.

        Args:
            query (str): Query.
            vectors (List[float]): Query vector.
            limit (int, optional): Number of results to return. Defaults to 5.
            filters (Dict, optional): Filters to apply to the search. Defaults to None.

        Returns:
            list: Search results.
        """
        filter_conditions = []
        filter_params = []

        if filters:
            for k, v in filters.items():
                filter_conditions.append("payload->>%s = %s")
                filter_params.extend([k, str(v)])

        filter_clause = "WHERE " + " AND ".join(filter_conditions) if filter_conditions else ""

        self.cur.execute(
            f"""
            SELECT id, vector <=> %s::vector AS distance, payload
            FROM {self.collection_name}
            {filter_clause}
            ORDER BY distance
            LIMIT %s
        """,
            (vectors, *filter_params, limit),
        )

        results = self.cur.fetchall()
        return [OutputData(id=str(r[0]), score=float(r[1]), payload=r[2]) for r in results]

    def delete(self, vector_id):
        """
        Delete a vector by ID.

        Args:
            vector_id (str): ID of the vector to delete.
        """
        self.cur.execute(f"DELETE FROM {self.collection_name} WHERE id = %s", (vector_id,))
        self.conn.commit()

    def update(self, vector_id, vector=None, payload=None):
        """
        Update a vector and its payload.

        Args:
            vector_id (str): ID of the vector to update.
            vector (List[float], optional): Updated vector.
            payload (Dict, optional): Updated payload.
        """
        if vector:
            self.cur.execute(
                f"UPDATE {self.collection_name} SET vector = %s WHERE id = %s",
                (vector, vector_id),
            )
        if payload:
            # Handle JSON serialization based on psycopg version
            if PSYCOPG_VERSION == 3:
                # psycopg3 uses psycopg.types.json.Json
                self.cur.execute(
                    f"UPDATE {self.collection_name} SET payload = %s WHERE id = %s",
                    (Json(payload), vector_id),
                )
            else:
                # psycopg2 uses psycopg2.extras.Json
                self.cur.execute(
                    f"UPDATE {self.collection_name} SET payload = %s WHERE id = %s",
                    (psycopg2.extras.Json(payload), vector_id),
                )
        self.conn.commit()

    def get(self, vector_id) -> OutputData:
        """
        Retrieve a vector by ID.

        Args:
            vector_id (str): ID of the vector to retrieve.

        Returns:
            OutputData: Retrieved vector.
        """
        self.cur.execute(
            f"SELECT id, vector, payload FROM {self.collection_name} WHERE id = %s",
            (vector_id,),
        )
        result = self.cur.fetchone()
        if not result:
            return None
        return OutputData(id=str(result[0]), score=None, payload=result[2])

    def list_cols(self) -> List[str]:
        """
        List all collections.

        Returns:
            List[str]: List of collection names.
        """
        self.cur.execute("SELECT table_name FROM information_schema.tables WHERE table_schema = 'public'")
        return [row[0] for row in self.cur.fetchall()]

    def delete_col(self):
        """Delete a collection."""
        self.cur.execute(f"DROP TABLE IF EXISTS {self.collection_name}")
        self.conn.commit()

    def col_info(self):
        """
        Get information about a collection.

        Returns:
            Dict[str, Any]: Collection information.
        """
        self.cur.execute(
            f"""
            SELECT 
                table_name, 
                (SELECT COUNT(*) FROM {self.collection_name}) as row_count,
                (SELECT pg_size_pretty(pg_total_relation_size('{self.collection_name}'))) as total_size
            FROM information_schema.tables 
            WHERE table_schema = 'public' AND table_name = %s
        """,
            (self.collection_name,),
        )
        result = self.cur.fetchone()
        return {"name": result[0], "count": result[1], "size": result[2]}

    def list(self, filters=None, limit=100):
        """
        List all vectors in a collection.

        Args:
            filters (Dict, optional): Filters to apply to the list.
            limit (int, optional): Number of vectors to return. Defaults to 100.

        Returns:
            List[OutputData]: List of vectors.
        """
        filter_conditions = []
        filter_params = []

        if filters:
            for k, v in filters.items():
                filter_conditions.append("payload->>%s = %s")
                filter_params.extend([k, str(v)])

        filter_clause = "WHERE " + " AND ".join(filter_conditions) if filter_conditions else ""

        query = f"""
            SELECT id, vector, payload
            FROM {self.collection_name}
            {filter_clause}
            LIMIT %s
        """

        self.cur.execute(query, (*filter_params, limit))

        results = self.cur.fetchall()
        return [[OutputData(id=str(r[0]), score=None, payload=r[2]) for r in results]]

    def __del__(self):
        """
        Close the database connection when the object is deleted.
        """
        if hasattr(self, "cur"):
            self.cur.close()
        if hasattr(self, "conn"):
            if hasattr(self, "connection_pool") and self.connection_pool is not None:
                # Return connection to pool instead of closing it
                self.connection_pool.putconn(self.conn)
            else:
                # Close the connection directly
                self.conn.close()

    def reset(self):
        """Reset the index by deleting and recreating it."""
        logger.warning(f"Resetting index {self.collection_name}...")
        self.delete_col()
        self.create_col(self.embedding_model_dims)

File: mem0-ts/src/oss/src/vector_stores/pgvector.ts
import type { Client as ClientType, ClientConfig } from "pg";
import pkg from "pg";
const { Client, escapeIdentifier } = pkg;
import { VectorStore } from "./base";
import { SearchFilters, VectorStoreConfig, VectorStoreResult } from "../types";

const SAFE_IDENTIFIER_RE = /^[a-zA-Z_][a-zA-Z0-9_]{0,127}$/;

function validateIdentifier(
  name: string,
  label: string = "identifier",
): string {
  if (!SAFE_IDENTIFIER_RE.test(name)) {
    throw new Error(
      `Invalid ${label} '${name}': only letters, digits, and underscores are allowed, ` +
        `must start with a letter or underscore, and be at most 128 characters.`,
    );
  }
  return name;
}

function escapeFilterKey(key: string): string {
  if (!SAFE_IDENTIFIER_RE.test(key)) {
    throw new Error(
      `Invalid filter key '${key}': only letters, digits, and underscores are allowed.`,
    );
  }
  return key;
}

interface FilterResult {
  conditions: string[];
  values: any[];
  paramIndex: number;
}

const OPERATOR_SQL_MAP: Record<string, { template: string; numeric: boolean }> =
  {
    eq: { template: "payload->>'%KEY%' = $%IDX%", numeric: false },
    ne: { template: "payload->>'%KEY%' != $%IDX%", numeric: false },
    gt: { template: "(payload->>'%KEY%')::numeric > $%IDX%", numeric: true },
    gte: { template: "(payload->>'%KEY%')::numeric >= $%IDX%", numeric: true },
    lt: { template: "(payload->>'%KEY%')::numeric < $%IDX%", numeric: true },
    lte: { template: "(payload->>'%KEY%')::numeric <= $%IDX%", numeric: true },
    in: { template: "payload->>'%KEY%' = ANY($%IDX%::text[])", numeric: false },
    nin: {
      template: "NOT (payload->>'%KEY%' = ANY($%IDX%::text[]))",
      numeric: false,
    },
    contains: {
      template: "payload->>'%KEY%' LIKE $%IDX% ESCAPE '\\'",
      numeric: false,
    },
    icontains: {
      template: "payload->>'%KEY%' ILIKE $%IDX% ESCAPE '\\'",
      numeric: false,
    },
  };

export function buildFilterConditions(
  filters: Record<string, any> | undefined,
  startIndex: number,
): FilterResult {
  const conditions: string[] = [];
  const values: any[] = [];
  let paramIndex = startIndex;

  if (!filters) {
    return { conditions, values, paramIndex };
  }

  for (const [key, value] of Object.entries(filters)) {
    if (key === "$or") {
      const orGroups: string[] = [];
      for (const orFilter of value as Record<string, any>[]) {
        const sub = buildFilterConditions(orFilter, paramIndex);
        if (sub.conditions.length > 0) {
          orGroups.push("(" + sub.conditions.join(" AND ") + ")");
          values.push(...sub.values);
          paramIndex = sub.paramIndex;
        }
      }
      if (orGroups.length > 0) {
        conditions.push("(" + orGroups.join(" OR ") + ")");
      }
      continue;
    }

    if (key === "$not") {
      const notGroups: string[] = [];
      for (const notFilter of value as Record<string, any>[]) {
        const sub = buildFilterConditions(notFilter, paramIndex);
        if (sub.conditions.length > 0) {
          notGroups.push("(" + sub.conditions.join(" AND ") + ")");
          values.push(...sub.values);
          paramIndex = sub.paramIndex;
        }
      }
      if (notGroups.length > 0) {
        conditions.push("NOT (" + notGroups.join(" OR ") + ")");
      }
      continue;
    }

    const safeKey = escapeFilterKey(key);

    if (value === "*") {
      conditions.push(`payload ? $${paramIndex}`);
      values.push(key);
      paramIndex++;
      continue;
    }

    if (typeof value === "object" && value !== null && !Array.isArray(value)) {
      for (const [op, opValue] of Object.entries(value)) {
        const mapping = OPERATOR_SQL_MAP[op];
        if (!mapping) {
          throw new Error(`Unsupported filter operator: ${op}`);
        }
        const clause = mapping.template
          .replace("%KEY%", safeKey)
          .replace("%IDX%", String(paramIndex));
        conditions.push(clause);

        if (op === "in" || op === "nin") {
          values.push((opValue as any[]).map(String));
        } else if (op === "contains" || op === "icontains") {
          const escaped = String(opValue)
            .replace(/\\/g, "\\\\")
            .replace(/%/g, "\\%")
            .replace(/_/g, "\\_");
          values.push(`%${escaped}%`);
        } else if (mapping.numeric) {
          values.push(Number(opValue));
        } else {
          values.push(String(opValue));
        }
        paramIndex++;
      }
    } else if (Array.isArray(value)) {
      conditions.push(`payload->>'${safeKey}' = ANY($${paramIndex}::text[])`);
      values.push(value.map(String));
      paramIndex++;
    } else {
      conditions.push(`payload->>'${safeKey}' = $${paramIndex}`);
      if (typeof value === "boolean") {
        values.push(JSON.stringify(value));
      } else {
        values.push(String(value));
      }
      paramIndex++;
    }
  }

  return { conditions, values, paramIndex };
}

interface PGVectorConfig extends VectorStoreConfig {
  dbname?: string;
  user?: string;
  password?: string;
  host?: string;
  port?: number;
  connectionString?: string;
  ssl?: ClientConfig["ssl"];
  embeddingModelDims: number;
  diskann?: boolean;
  hnsw?: boolean;
}

function getConnectionString(config: PGVectorConfig): string | undefined {
  return config.connectionString?.trim() || undefined;
}

function validateConnectionConfig(config: PGVectorConfig): void {
  if (getConnectionString(config)) {
    return;
  }

  const missingFields = ["user", "password", "host", "port"].filter(
    (field) => config[field as keyof PGVectorConfig] === undefined,
  );

  if (missingFields.length > 0) {
    throw new Error(
      `PGVector requires either connectionString or ${missingFields.join(", ")}`,
    );
  }
}

function buildClientConfig(
  config: PGVectorConfig,
  database?: string,
): ClientConfig {
  const connectionString = getConnectionString(config);
  if (connectionString) {
    return {
      connectionString,
      ...(config.ssl !== undefined ? { ssl: config.ssl } : {}),
    };
  }

  return {
    database,
    user: config.user,
    password: config.password,
    host: config.host,
    port: config.port,
    ...(config.ssl !== undefined ? { ssl: config.ssl } : {}),
  };
}

export class PGVector implements VectorStore {
  private client: ClientType;
  private collectionName: string;
  private useDiskann: boolean;
  private useHnsw: boolean;
  private readonly dbName: string;
  private readonly useDirectConnection: boolean;
  private config: PGVectorConfig;
  private _initPromise?: Promise<void>;

  constructor(config: PGVectorConfig) {
    validateConnectionConfig(config);
    this.collectionName = validateIdentifier(
      config.collectionName || "memories",
      "collectionName",
    );
    this.useDiskann = config.diskann || false;
    this.useHnsw = config.hnsw || false;
    this.useDirectConnection = !!getConnectionString(config);
    this.dbName = this.useDirectConnection
      ? ""
      : validateIdentifier(config.dbname || "vector_store", "dbname");
    this.config = config;

    this.client = new Client(
      buildClientConfig(
        config,
        this.useDirectConnection ? undefined : "postgres",
      ),
    );
    this.initialize().catch(console.error);
  }

  private col(): string {
    return escapeIdentifier(this.collectionName);
  }

  async initialize(): Promise<void> {
    if (!this._initPromise) {
      this._initPromise = this._doInitialize();
    }
    return this._initPromise;
  }

  private async _doInitialize(): Promise<void> {
    try {
      await this.client.connect();

      if (!this.useDirectConnection) {
        const dbExists = await this.checkDatabaseExists(this.dbName);
        if (!dbExists) {
          await this.createDatabase(this.dbName);
        }

        await this.client.end();

        this.client = new Client(buildClientConfig(this.config, this.dbName));
        await this.client.connect();
      }

      await this.client.query("CREATE EXTENSION IF NOT EXISTS vector");

      await this.client.query(`
        CREATE TABLE IF NOT EXISTS memory_migrations (
          id SERIAL PRIMARY KEY,
          user_id TEXT NOT NULL UNIQUE
        )
      `);

      const collections = await this.listCols();
      if (!collections.includes(this.collectionName)) {
        await this.createCol(this.config.embeddingModelDims);
      }
    } catch (error) {
      console.error("Error during initialization:", error);
      throw error;
    }
  }

  private async checkDatabaseExists(dbName: string): Promise<boolean> {
    const result = await this.client.query(
      "SELECT 1 FROM pg_database WHERE datname = $1",
      [dbName],
    );
    return result.rows.length > 0;
  }

  private async createDatabase(dbName: string): Promise<void> {
    await this.client.query(`CREATE DATABASE ${escapeIdentifier(dbName)}`);
  }

  private async createCol(embeddingModelDims: number): Promise<void> {
    const dims = Math.floor(embeddingModelDims);
    await this.client.query(`
      CREATE TABLE IF NOT EXISTS ${this.col()} (
        id UUID PRIMARY KEY,
        vector vector(${dims}),
        payload JSONB
      );
    `);

    if (this.useDiskann && embeddingModelDims < 2000) {
      try {
        const result = await this.client.query(
          "SELECT * FROM pg_extension WHERE extname = 'vectorscale'",
        );
        if (result.rows.length > 0) {
          await this.client.query(`
            CREATE INDEX IF NOT EXISTS ${escapeIdentifier(this.collectionName + "_diskann_idx")}
            ON ${this.col()}
            USING diskann (vector);
          `);
        }
      } catch (error) {
        console.warn("DiskANN index creation failed:", error);
      }
    } else if (this.useHnsw) {
      try {
        await this.client.query(`
          CREATE INDEX IF NOT EXISTS ${escapeIdentifier(this.collectionName + "_hnsw_idx")}
          ON ${this.col()}
          USING hnsw (vector vector_cosine_ops);
        `);
      } catch (error) {
        console.warn("HNSW index creation failed:", error);
      }
    }
  }

  async insert(
    vectors: number[][],
    ids: string[],
    payloads: Record<string, any>[],
  ): Promise<void> {
    const values = vectors.map((vector, i) => ({
      id: ids[i],
      vector: `[${vector.join(",")}]`,
      payload: payloads[i],
    }));

    const query = `
      INSERT INTO ${this.col()} (id, vector, payload)
      VALUES ($1, $2::vector, $3::jsonb)
    `;

    await Promise.all(
      values.map((value) =>
        this.client.query(query, [value.id, value.vector, value.payload]),
      ),
    );
  }

  async keywordSearch(
    query: string,
    topK: number = 5,
    filters?: SearchFilters,
  ): Promise<VectorStoreResult[] | null> {
    try {
      const {
        conditions,
        values,
        paramIndex: _,
      } = buildFilterConditions(filters, 3);
      const filterValues: any[] = [query, topK, ...values];

      const filterClause =
        conditions.length > 0 ? "AND " + conditions.join(" AND ") : "";

      const searchQuery = `
        SELECT id, ts_rank_cd(to_tsvector('simple', payload->>'textLemmatized'), plainto_tsquery('simple', $1)) AS score, payload
        FROM ${this.col()}
        WHERE to_tsvector('simple', payload->>'textLemmatized') @@ plainto_tsquery('simple', $1)
        ${filterClause}
        ORDER BY score DESC
        LIMIT $2
      `;

      const result = await this.client.query(searchQuery, filterValues);

      return result.rows.map((row) => ({
        id: row.id,
        payload: row.payload,
        score: row.score,
      }));
    } catch (error) {
      console.error("Error during keyword search:", error);
      return null;
    }
  }

  async search(
    query: number[],
    topK: number = 5,
    filters?: SearchFilters,
  ): Promise<VectorStoreResult[]> {
    const queryVector = `[${query.join(",")}]`;
    const {
      conditions,
      values,
      paramIndex: _,
    } = buildFilterConditions(filters, 3);
    const filterValues: any[] = [queryVector, topK, ...values];

    const filterClause =
      conditions.length > 0 ? "WHERE " + conditions.join(" AND ") : "";

    const searchQuery = `
      SELECT id, vector <=> $1::vector AS distance, payload
      FROM ${this.col()}
      ${filterClause}
      ORDER BY distance
      LIMIT $2
    `;

    const result = await this.client.query(searchQuery, filterValues);

    return result.rows.map((row) => ({
      id: row.id,
      payload: row.payload,
      score: Math.max(0, Math.min(1, 1 - Number(row.distance))),
    }));
  }

  async get(vectorId: string): Promise<VectorStoreResult | null> {
    const result = await this.client.query(
      `SELECT id, payload FROM ${this.col()} WHERE id = $1`,
      [vectorId],
    );

    if (result.rows.length === 0) return null;

    return {
      id: result.rows[0].id,
      payload: result.rows[0].payload,
    };
  }

  async update(
    vectorId: string,
    vector: number[],
    payload: Record<string, any>,
  ): Promise<void> {
    const vectorStr = `[${vector.join(",")}]`;
    await this.client.query(
      `
      UPDATE ${this.col()}
      SET vector = $1::vector, payload = $2::jsonb
      WHERE id = $3
      `,
      [vectorStr, payload, vectorId],
    );
  }

  async delete(vectorId: string): Promise<void> {
    await this.client.query(`DELETE FROM ${this.col()} WHERE id = $1`, [
      vectorId,
    ]);
  }

  async deleteCol(): Promise<void> {
    await this.client.query(`DROP TABLE IF EXISTS ${this.col()}`);
  }

  private async listCols(): Promise<string[]> {
    const result = await this.client.query(`
      SELECT table_name
      FROM information_schema.tables
      WHERE table_schema = 'public'
    `);
    return result.rows.map((row) => row.table_name);
  }

  async list(
    filters?: SearchFilters,
    topK: number = 100,
  ): Promise<[VectorStoreResult[], number]> {
    const {
      conditions,
      values: filterValues,
      paramIndex,
    } = buildFilterConditions(filters, 1);

    const filterClause =
      conditions.length > 0 ? "WHERE " + conditions.join(" AND ") : "";

    const listQuery = `
      SELECT id, payload
      FROM ${this.col()}
      ${filterClause}
      LIMIT $${paramIndex}
    `;

    const countQuery = `
      SELECT COUNT(*)
      FROM ${this.col()}
      ${filterClause}
    `;

    const listValues = [...filterValues, topK];

    const [listResult, countResult] = await Promise.all([
      this.client.query(listQuery, listValues),
      this.client.query(countQuery, filterValues),
    ]);

    const results = listResult.rows.map((row) => ({
      id: row.id,
      payload: row.payload,
    }));

    return [results, parseInt(countResult.rows[0].count)];
  }

  async close(): Promise<void> {
    await this.client.end();
  }

  async getUserId(): Promise<string> {
    const result = await this.client.query(
      "SELECT user_id FROM memory_migrations LIMIT 1",
    );

    if (result.rows.length > 0) {
      return result.rows[0].user_id;
    }

    // Generate a random user_id if none exists
    const randomUserId =
      Math.random().toString(36).substring(2, 15) +
      Math.random().toString(36).substring(2, 15);
    await this.client.query(
      "INSERT INTO memory_migrations (user_id) VALUES ($1)",
      [randomUserId],
    );
    return randomUserId;
  }

  async setUserId(userId: string): Promise<void> {
    await this.client.query("DELETE FROM memory_migrations");
    await this.client.query(
      "INSERT INTO memory_migrations (user_id) VALUES ($1)",
      [userId],
    );
  }
}
